xmtp_mls/src/groups/device_sync/backup/export_stream.rs

```diff
@@ -0,0 +1,139 @@
+use super::BackupOptions;
+use crate::XmtpOpenMlsProvider;
+use futures::{Stream, StreamExt};
+use std::{marker::PhantomData, pin::Pin, sync::Arc, task::Poll};
+use xmtp_proto::xmtp::device_sync::{
+    consent_backup::ConsentSave, group_backup::GroupSave, message_backup::GroupMessageSave,
+    BackupElement, BackupElementSelection,
+};
+
+pub(crate) mod consent_save;
+pub(crate) mod group_save;
+pub(crate) mod message_save;
+
+type BackupInputStream = Pin<Box<dyn Stream<Item = Vec<BackupElement>> + Send>>;
+
+/// A stream that curates a collection of streams for backup.
+pub(super) struct BatchExportStream {
+    pub(super) buffer: Vec<BackupElement>,
+    pub(super) input_streams: Vec<BackupInputStream>,
+}
+
+impl BatchExportStream {
+    pub(super) fn new(opts: &BackupOptions, provider: &Arc<XmtpOpenMlsProvider>) -> Self {
+        let input_streams = opts
+            .elements
+            .iter()
+            .flat_map(|&e| match e {
+                BackupElementSelection::Consent => {
+                    vec![BackupRecordStreamer::<ConsentSave>::new_stream(
+                        provider, opts,
+                    )]
+                }
+                BackupElementSelection::Messages => vec![
+                    // Order matters here. Don't put groups before messages.
```

Comment:
what happens if groups come before messages?

Reply:
The insertion order is opposite the order here, and inserting groups before messages will cause a FK constraint error. I could probably reverse the iterator when building the streams to make that more clear..

Reply:
Fixed.